File: frontend/src/lib/api/report-publications.ts
/** Report publications API client (ADR-186). */

import { apiFetch } from "./core";

export interface ReportPublicationResponse {
  id: string;
  workspace_id: string;
  period_yyyymm: string;
  artifact_id: number;
  published_at: string;
  published_by: string;
  immutable_hash: string;
  unpublished_at: string | null;
}

export interface ReportPublicationListResponse {
  items: ReportPublicationResponse[];
}

/** Normaliza `period` (`2026-04` ou `202604`) para o formato `YYYYMM` exigido pela API. */
export function normalizePeriodYyyymm(period: string | null | undefined): string | null {
  if (!period) return null;
  const digits = period.replace(/\D/g, "");
  return digits.length === 6 ? digits : null;
}

export async function getActivePublication(
  workspaceId: string,
  periodYyyymm: string,
): Promise<ReportPublicationResponse | null> {
  return apiFetch(
    `/workspaces/${workspaceId}/reports/${periodYyyymm}/publication`,
  );
}

export async function listPublications(
  workspaceId: string,
): Promise<ReportPublicationListResponse> {
  return apiFetch(`/workspaces/${workspaceId}/reports/publications`);
}

export async function publishMonth(
  workspaceId: string,
  periodYyyymm: string,
  artifactId: number,
): Promise<ReportPublicationResponse> {
  return apiFetch(`/workspaces/${workspaceId}/reports/${periodYyyymm}/publish`, {
    method: "POST",
    body: JSON.stringify({ artifact_id: artifactId }),
    headers: { "Content-Type": "application/json" },
  });
}

export async function unpublishMonth(
  workspaceId: string,
  periodYyyymm: string,
): Promise<void> {
  await apiFetch(
    `/workspaces/${workspaceId}/reports/${periodYyyymm}/publish`,
    { method: "DELETE" },
  );
}

File: backend/app/api/report_publications.py
"""Report publications API — ADR-186 (mês fechado imutável)."""

from __future__ import annotations

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.core.database import get_db
from backend.app.core.deps import get_current_user
from backend.app.core.tenancy import get_current_workspace, require_write_role
from backend.app.models.user import User
from backend.app.models.workspace import Workspace
from backend.app.schemas.report_publication import (
    ReportPublicationCreate,
    ReportPublicationListResponse,
    ReportPublicationResponse,
    to_response,
)
from backend.app.services import report_publication as report_publication_service

router = APIRouter(
    prefix="/workspaces/{workspace_id}/reports",
    tags=["report-publications"],
)


def _actor_id(user: User) -> str:
    return f"user:{user.id}"


@router.post(
    "/{period_yyyymm}/publish",
    response_model=ReportPublicationResponse,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_write_role)],
)
async def publish_report(
    period_yyyymm: str,
    payload: ReportPublicationCreate,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ReportPublicationResponse:
    publication = await report_publication_service.publish_month(
        workspace.id,
        period_yyyymm,
        payload.artifact_id,
        actor=_actor_id(user),
        db=db,
    )
    await db.commit()
    return to_response(publication)


@router.delete(
    "/{period_yyyymm}/publish",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_write_role)],
)
async def unpublish_report(
    period_yyyymm: str,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    await report_publication_service.unpublish_month(
        workspace.id,
        period_yyyymm,
        actor=_actor_id(user),
        db=db,
    )
    await db.commit()
    return None


@router.get(
    "/publications",
    response_model=ReportPublicationListResponse,
)
async def list_publications(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
) -> ReportPublicationListResponse:
    publications = await report_publication_service.list_publications(workspace.id, db=db)
    return ReportPublicationListResponse(items=[to_response(p) for p in publications])


@router.get(
    "/{period_yyyymm}/publication",
    response_model=ReportPublicationResponse | None,
)
async def get_active_publication(
    period_yyyymm: str,
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
) -> ReportPublicationResponse | None:
    """Publicação viva do (workspace, period), ou ``null`` se mês está aberto."""
    publication = await report_publication_service.get_active_publication(
        workspace.id, period_yyyymm, db=db
    )
    if publication is None:
        return None
    return to_response(publication)
